hw2_pandas/hw2_pandas.py

```diff
@@ -0,0 +1,407 @@
+# properties
+#     - this is logical - Total is the sum of all combat properties.
+# 3. Generation - does not correlate with other characteristics.
+# 4. Legendaries - significantly correlates with combat characteristics,
+# which is also logical.
+#
+# ### Conclusion
+# It is more convenient to do the initial analysis of the data using the
+# pandas_profiling library.
+#
+# This will allow to see explicit dependencies and, if necessary, examine
+# them in more details. Data analysis "manually" using Pandas on this data
+# set turned out to be poorly informative, since in most cases dependencies
+# between variables are weakly expressed.
+
+# In[16]:
+
+
+###############################################################################
+# Task 4 - Extra task (extra 25 points)
+###############################################################################
+
+# reading data from .gff file to pandas.DataFrame
+def read_gff(path):
+    # line format has view:
+    #   Reference_10\t barrnap:0.9\t rRNA\t 67122\t 68652\t 0.0\t +\t .\t
+    #   Name=16S_rRNA;product=16S ribosomal RNA
+    return pd.read_csv(path, sep='\t', skiprows=1,
+                       names=['chromosome', 'source', 'type', 'start', 'end',
+                              'score', 'strand', 'phase', 'attributes'])
+
+
+df_gff = read_gff('rrna_annotation.gff')
+df_gff
+
+
+# In[17]:
+
+
+# reading data from .bed6 file to pandas.DataFrame
+def read_bed6(path):
+    # line format has view:
+    #   Reference_1/t 197681/t 200286\t NODE_1445_length_2603_cov_1135/t
+    #   593799/t  41/t -
+    return pd.read_csv(path, sep='\t',
+                       names=['chromosome', 'start', 'end', 'name',
+                              'score', 'strand'])
+
+
+df_bed = read_bed6('alignment.bed')
+df_bed
+
+
+# In[18]:
+
+
+# modifying of values in 'attributes' column
+
+def make_attr_shoter(attr_str):
+    m = re.search('product=(.+?) ', attr_str)
+    if m:
+        return m.group(1)
+    return None
+
+
+df_gff = read_gff('rrna_annotation.gff')
+df_gff['attributes'] = df_gff['attributes'].apply(make_attr_shoter)
+df_gff
+
+
+# In[19]:
+
+
+# Make a table where for each chromosome (in fact, these are not chromosomes,
+# but reference genomes) shows the amount of rRNA of each type.
+# We can build a barplot that displays this data
+
+df_gff['attributes'].value_counts()
+# There are only 3 types - 23S, 16S and 5S
+
+
+# In[20]:
+
+
+# Look through types of chromosome
+df_gff['chromosome'].value_counts().sort_index()
+# There are 26 types of chromosome: [Reference_1 .. Reference_26]
+
+
+# In[21]:
+
+
+# make new DF with counting of RNA for each type
+df_counts = pd.DataFrame(columns=['chromosome', '5S', '16S', '23S'])
+for i in range(26):
+    ref = 'Reference_' + str(i + 1)
+    cnts = df_gff.query('chromosome == @ref')['attributes'].value_counts()
+    df_counts.loc[i] = [ref, cnts.get('5S', 0), cnts.get('16S', 0),
+                        cnts.get('23S', 0)]
+
+df_counts
```

Comment:
Можно было сделать без цикла, например, вот так
```suggestion
df = df_gff.groupby(["chromosome", "attributes"], as_index=False).agg({"type": "count"})
sns.barplot(y="chromosome", x="type", hue="attributes", data=df, orient="h")
```